### PR TITLE
CDI-494 Events - clarify a wildcard type is not an unresolvable type variable

### PR DESCRIPTION
**Pull request by mkouba**
* https://issues.jboss.org/browse/CDI-494
* http://lists.jboss.org/pipermail/cdi-dev/2016-February/007789.html

### CODE REVIEW

**Comment by manovotn** [comment]
+1 from me

Though I was thinking a  tiny code snippet showing some of the possibilities it gives would not hurt, WDYT?

**Comment by mkouba** [comment]
Sure, this would enable e.g. the following code to be portable: 
```java
@Inject Event<List<?>> event;

void process() {
  List<String> stringList = new ArrayList<>();
  event.fire(stringList);
  List<Integer> intList = new ArrayList<>();
  event.fire(intList);
}

void observeAllLists(@Observes List<?> anyList) {
  ...
}
```
Note that in this example the runtime type of events is `ArrayList<E>` and `E` could be resolved to `?`.

**Comment by manovotn** [comment]
I meant the snippet for spec, as talking wildcards can get a bit wild at time and not every reader will grasp this without a simple sample. On the other hand it might be unnecessary, we can wait and see what others think :)

File: spec/src/main/asciidoc/core/events.asciidoc [hunk 1]
```diff
@@ -26,7 +26,7 @@ An observer method will be notified of an event if the event object is assignabl
 An event object is an instance of a concrete Java class with no unresolvable type variables.
 The _event types_ of the event include all superclasses and interfaces of the runtime class of the event object.
 
-An event type may not contain an unresolvable type variable.
+An event type may not contain an unresolvable type variable. A wildcard type is not considered an unresolvable type variable.
```
Comment (antoinesd) [inline spec/src/main/asciidoc/core/events.asciidoc]: Be careful here. In JLS Wildcard type are unbound (<?>) or bound (<? extends Number>) are you sure you mean both here?

**Comment by antoinesd** [comment]
+1 for the PR if we talk about unbounded wildcard. Bounded wildcard would be nice, but I don't see how we could define them as resolvable type when all type resolution in the spec don't accept it.

File: spec/src/main/asciidoc/core/events.asciidoc (hunk 1 above)
Comment (mkouba) [inline spec/src/main/asciidoc/core/events.asciidoc]: Yes, for both. `<?>` translates to `<? extends Object>`.

**Comment by mkouba** [comment]
> ...when all type resolution in the spec don't accept it

What exactly do you mean?

**Comment by antoinesd** [comment]
All type resolution section in the spec, only mention unbounded wildcard like here:
http://docs.jboss.org/cdi/spec/1.2/cdi-spec.html#assignable_parameters
here:
http://docs.jboss.org/cdi/spec/1.2/cdi-spec.html#bean_metadata

**Comment by mkouba** [comment]
Hm, I think the spec always mentions the wildcard bounds... And as I wrote above - the wildcard `? extends Object` is equivalent to the unbounded wildcard `?` (JLS "4.5.1 Type Arguments of Parameterized Types").

File: spec/src/main/asciidoc/core/events.asciidoc (hunk 1 above)
Comment (antoinesd) [inline spec/src/main/asciidoc/core/events.asciidoc]: `<? extends Object>` means any type while `<? extends Number>` means any type extending number, I think the restriction makes a lot of difference here, hence the therm "bounded" in JLS.

Now, if there are no technical problem at the impl level to support bounded wildcard, why not...

**Comment by antoinesd** [comment]
We're saying the same: the spec only accept (and mention) **unbounded wildcards**. So I only pointed that allowing **bounded wildcards** should be done carefully (even if events are not the same than typesafe resolution).

**Comment by mkouba** [comment]
The same except the other way around ;-)  The spec only mentions bounded wildcards...

**Comment by antoinesd** [comment]
ok, my mistake :-).
Does the mention below (in 5.2.3) could be an inspiration for this PR?

> Any legal bean type, as defined in Legal bean types may be the required type of an injection point. Furthermore, **the required type of an injection point may contain a wildcard type parameter**. However, a type variable is not a legal injection point type.

**Comment by mkouba** [comment]
See also https://github.com/mkouba/event-wildcard-test

**Comment by slaskawi** [comment]
+1 for this feature - we need it in Infinispan (oftentimes when we fire event we don't know their types - they are use driven).

**Comment by antoinesd** [comment]
+1 for me as well

**Comment by antoinesd** [comment]
@slaskawi very interesting to have feedback from users. thx.
Unless we missed a big issue with this approach I'm +1 for it

**Comment by johnament** [comment]
+1 from me as well.

**Pull request merged**